### PR TITLE
UCP/RMA/TEST: Add new protocols for ucp_get operations

### PR DESCRIPTION
**Pull request by yosefe**
# Why
Implement ucp_get* APIs over new protocols data-structures

# How 
* Implement protocol progress for get/am, get/offload.bcopy, get/offload/zcopy
* In ucp_get_nbx() [common function for all ucp_get* APIs] ,check if protocols are enabled, and if yes initialze and start the send request with new protocol progress functions

### CODE REVIEW

**Comment by shamisp** [comment]
@yosefe can you please add why what  ?

File: src/ucp/proto/proto_common.h [hunk 1]
```diff
@@ -11,14 +11,22 @@
...
 typedef enum {
     UCP_PROTO_COMMON_INIT_FLAG_SEND_ZCOPY    = UCS_BIT(0), /* Send buffer is used by
                                                               zero-copy operations */
     UCP_PROTO_COMMON_INIT_FLAG_RECV_ZCOPY    = UCS_BIT(1), /* Receive side is not
                                                               doing memory copy */
     UCP_PROTO_COMMON_INIT_FLAG_REMOTE_ACCESS = UCS_BIT(2), /* One-sided remote access */
-    UCP_PROTO_COMMON_INIT_FLAG_MEM_TYPE      = UCS_BIT(3)  /* Protocol supports
+    UCP_PROTO_COMMON_INIT_FLAG_MEM_TYPE      = UCS_BIT(3), /* Protocol supports
                                                               non-host buffers */
+    UCP_PROTO_COMMON_INIT_FLAG_RESPONSE      = UCS_BIT(4), /* Sender side has to get
```
Comment (shamisp) [inline src/ucp/proto/proto_common.h:26]: How it is different from send zcopy ?

File: src/ucp/proto/proto_common.h [hunk 2]
```diff
@@ -11,14 +11,22 @@
...
+                                                              a reply before can
+                                                              complete the request */
+    UCP_PROTO_COMMON_INIT_FLAG_HDR_ONLY      = UCS_BIT(5)  /* Header only, no payload */
```
Comment (shamisp) [inline src/ucp/proto/proto_common.h:29]: I'm not sure what header only means.

File: src/ucp/proto/proto_common.h [hunk 3]
```diff
@@ -30,7 +38,9 @@ typedef struct {
     double                  overhead;      /* protocol overhead */
     size_t                  cfg_thresh;    /* user-configured threshold */
     unsigned                cfg_priority;  /* user configuration priority */
-    size_t                  fragsz_offset; /* offset of maximal fragment
+    ptrdiff_t               min_sz_offset; /* offset of minimal size in
```
Comment (shamisp) [inline src/ucp/proto/proto_common.h:41]: Minimal size of what ?

File: src/ucp/rma/get_am.c [hunk 4]
```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
```
Comment (shamisp) [inline src/ucp/rma/get_am.c:2]: Are you sure the affiliation is correct ?

**Review by shamisp — CHANGES_REQUESTED** [review]

File: src/ucp/proto/proto_common.h (hunk 1 above)
Comment (yosefe) [reply src/ucp/proto/proto_common.h:26]: It means the buck data transfer is don from target->initiator so no need to consider fragment size on sender side.

File: src/ucp/proto/proto_common.h [hunk 5]
```diff
@@ -30,7 +38,9 @@ typedef struct {
...
+                                              uct_iface_attr_t */
+    ptrdiff_t               fragsz_offset; /* offset of maximal fragment
```
Comment (brminich) [inline src/ucp/proto/proto_common.h:43]: minor: `fragsz_offset` name is not consistent with `min_sz_offset` (missing `_` after `frag`)

File: src/ucp/proto/proto_common.c [hunk 6]
```diff
@@ -283,7 +286,8 @@ ucp_proto_common_calc_completion(const ucp_proto_common_init_params_t *params,
     ucp_proto_perf_range_t *range =
             &params->super.caps->ranges[params->super.caps->num_ranges++];
 
-    if (perf_params->is_multi) {
+    if (perf_params->is_multi ||
+        (params->flags & UCP_PROTO_COMMON_INIT_FLAG_HDR_ONLY)) {
```
Comment (brminich) [inline src/ucp/proto/proto_common.c:290]: can header be bigger than single fragment?

File: src/ucp/proto/proto_common.h (hunk 1 above)
Comment (yosefe) [reply src/ucp/proto/proto_common.h]: in send_zcopy without flag FAST_COMPLETION would estimate the latency from sender to receiver, but in get_zcopy need to measure the latency until data arrives back (including both the request and the response)

File: src/ucp/proto/proto_common.h (hunk 2 above)
Comment (yosefe) [reply src/ucp/proto/proto_common.h]: replaced by other flag

File: src/ucp/proto/proto_common.h (hunk 3 above)
Comment (yosefe) [reply src/ucp/proto/proto_common.h]: added more docs

File: src/ucp/rma/get_am.c (hunk 4 above)
Comment (yosefe) [reply src/ucp/rma/get_am.c:2]: yes, we're still using this one

File: src/ucp/proto/proto_common.h (hunk 5 above)
Comment (yosefe) [reply src/ucp/proto/proto_common.h]: fixed

File: src/ucp/proto/proto_common.h [hunk 7]
```diff
@@ -38,10 +44,10 @@ typedef struct {
...
     size_t                  cfg_thresh;    /* user-configured threshold */
     unsigned                cfg_priority;  /* user configuration priority */
-    ptrdiff_t               min_sz_offset; /* offset of minimal size in
-                                              uct_iface_attr_t */
-    ptrdiff_t               fragsz_offset; /* offset of maximal fragment
-                                              size in uct_iface_attr_t */
+    ptrdiff_t               min_frag_offs; /* offset in uct_iface_attr_t of the
+                                              minimal size of a single fragment */
+    ptrdiff_t               max_frag_offs; /* offset in uct_iface_attr_t of the
+                                              maximal size of a single fragment */
```
Comment (brminich) [inline src/ucp/proto/proto_common.h:50]: 👍

**Review by brminich — APPROVED** [review]

File: src/ucp/proto/proto_common.c (hunk 6 above)
Comment (yosefe) [reply src/ucp/proto/proto_common.c]: fixed (in another commit)

**Comment by yosefe** [comment]
@shamisp @brminich can you pls look again?

File: src/ucp/proto/proto_common.c [hunk 8]
```diff
@@ -397,8 +407,11 @@ void ucp_proto_common_calc_perf(const ucp_proto_common_init_params_t *params,
     iface_attr = ucp_proto_common_get_iface_attr(&params->super,
                                                  perf_params->lane0);
     frag_size  = ucp_proto_get_iface_attr_field(iface_attr,
-                                                params->max_frag_offs, SIZE_MAX) -
-                 params->hdr_size;
+                                                params->max_frag_offs, SIZE_MAX);
+    if (!(params->flags & UCP_PROTO_COMMON_INIT_FLAG_RESPONSE)) {
+        /* if the data returns as a response, no need to subtract header size */
```
Comment (brminich) [inline src/ucp/proto/proto_common.c:412]: can some proto send data with another header?

Comment (yosefe) [reply src/ucp/proto/proto_common.c:412]: do you mean whether the response data may have a header as well, in addition to request?
currently no such protocol is expected so probably no need to add such support for now

Comment (shamisp) [reply src/ucp/proto/proto_common.c:412]: @yosefe maybe document this assumption in the code.

File: src/ucp/proto/proto_common.h [hunk 9]
```diff
@@ -11,14 +11,28 @@
...
+    /* Receive side is not doing memory copy */
+    UCP_PROTO_COMMON_INIT_FLAG_RECV_ZCOPY    = UCS_BIT(1),
+
+    /* One-sided remote access (implies RECV_ZCOPY) */
+    UCP_PROTO_COMMON_INIT_FLAG_REMOTE_ACCESS = UCS_BIT(2),
+
+    /* Protocol supports non-host buffers */
+    UCP_PROTO_COMMON_INIT_FLAG_MEM_TYPE      = UCS_BIT(3),
+
+    /* The data arrives to sender side, so latency is by send completion */
```
Comment (shamisp) [inline src/ucp/proto/proto_common.h:31]: s/sender/request initiator

**Review by brminich — APPROVED** [review]

**Comment by yosefe** [comment]
@shamisp updated documentation

**Review by shamisp — APPROVED** [review]

**Comment by yosefe** [comment]
MellanoxLab New passed but status is not updated

**Pull request merged**